### PR TITLE
Fix: Async component loader wrong path for 'async!...' imports

### LINKED ISSUE #47: require.ensure named-chunk from async-component-loader is unsafe for URL
# This is a bug report

# Describe the problem

Currently, async-component-loader names chunk with full path (something like /home/user/.../index.js)

Webpack then uses these names to emit files and then load them with relative paths. 

Webpack's code to load the chunk:

```js
var head = document.getElementsByTagName('head')[0];
/******/ 		var script = document.createElement('script');
/******/ 		script.type = 'text/javascript';
/******/ 		script.charset = 'utf-8';
/******/ 		script.async = true;
/******/ 		script.timeout = 120000;
/******/
/******/ 		if (__webpack_require__.nc) {
/******/ 			script.setAttribute("nonce", __webpack_require__.nc);
/******/ 		}
/******/ 		script.src = __webpack_require__.p + "" + ({"0":"polyfills","1":"route-home","2":"/home/khoa/web/tooth/frontend/routes/profile/index.js","3":"/home/khoa/web/tooth/frontend/components/SuggestionList/index.jsx","4":"route-profile"}[chunkId]||chunkId) + ".chunk." + {"0":"9226481e4a5e84a7ee24","1":"b5e4c77eba4ee1f85a5f","2":"30aa1718e1c2e1be2a10","3":"93378fcc5f457e450ce9","4":"df80d0c60dc51a254ff3"}[chunkId] + ".js";
```

The last line prettified:

```js
script.src =
  __webpack_require__.p +
  "" +
  ({
    "0": "polyfills",
    "1": "route-home",
    "2": "/home/khoa/web/tooth/frontend/routes/profile/index.js",
    "3": "/home/khoa/web/tooth/frontend/components/SuggestionList/index.jsx",
    "4": "route-profile"
  }[chunkId] || chunkId) +
  ".chunk." +
  {
    "0": "9226481e4a5e84a7ee24",
    "1": "b5e4c77eba4ee1f85a5f",
    "2": "30aa1718e1c2e1be2a10",
    "3": "93378fcc5f457e450ce9",
    "4": "df80d0c60dc51a254ff3"
  }[chunkId] +
  ".js";

```

This leads to 2 problems:

1. Unable to async load any components (except route chunks since they are named differently)

For ex, chunkId = 3 url (in the above example) is `http://home/khoa/web/tooth/frontend/components/SuggestionList/index.jsx.chunk.93378fcc5f457e450ce9.js` . Which is obviously results in a 404

2. Privacy

Leaking path to the public isn't something I want to do. 

# Suggestions

I suggest to generate chunk name using a quick hash function like sha-1 with relative path of the file in the project. This chunk name is persistent regarding the build path which is important for long-term caching.

P/S: Sorry in advance if my poor English skills annoyed you. 

---

Update:

I created a simple repo to reproduce this https://github.com/thangngoc89/preact-cli-47

Steps to repo:

- npm install
- npm start

(and look at the browser's console)

### PR DESCRIPTION
**Pull request by rkostrzewski**
Fixes #47

### CODE REVIEW

File: src/lib/async-component-loader.js [hunk 1]
```diff
@@ -6,7 +6,14 @@ module.exports.pitch = function(remainingRequest) {
 	this.cacheable && this.cacheable();
 	var query = loaderUtils.getOptions(this) || {};
 	var routeName = typeof query.name === 'function' ? query.name(this.resourcePath) : null;
-	var name = routeName !== null ? routeName : ('name' in query ? query.name : (query.formatName || String)(this.resourcePath));
+	var name
+	if(routeName !== null) {
```
Comment (marvinhagemeister) [inline src/lib/async-component-loader.js]: Nitpick: missing space after `if` keyword

File: package.json [hunk 2]
```diff
@@ -51,6 +52,8 @@
     "rules": {
       "no-console": 1,
       "no-empty": 0,
+      "semi": 2,
+      "keyword-spacing": 2,
```
Comment (rkostrzewski) [inline package.json]: @developit I took the liberty of adding those rules to eslint - please check if they're ok with you.
Semicolons are enforced (they were placed almost everywhere and not enforced) & spaces should be placed after keyword (no changes to files).

Comment (developit) [inline package.json]: both good with me, thanks a bunch!

File: src/lib/async-component-loader.js [hunk 3]
```diff
@@ -6,7 +6,14 @@ module.exports.pitch = function(remainingRequest) {
 	this.cacheable && this.cacheable();
 	var query = loaderUtils.getOptions(this) || {};
 	var routeName = typeof query.name === 'function' ? query.name(this.resourcePath) : null;
-	var name = routeName !== null ? routeName : ('name' in query ? query.name : (query.formatName || String)(this.resourcePath));
```
Comment (developit) [inline src/lib/async-component-loader.js]: I knew someone would undo this ~~lovely~~ upsetting ternary

**Comment by developit** [comment]
Are you able to shed light on what was messed up here?  I'm happy to merge the change, just curious if it was a logic error or something I'm not seeing.

**Comment by rkostrzewski** [comment]
Previously this.resourcePath (which is an absolute URI) was passed to require.ensure third param which is chunk name - this in turn led to chunk names having whole absolute path as name ^^.

**Pull request merged**